gramineproject/gramine
# Enable stack sanitization

## Description of the changes 

This enables stack sanitization in ASan, which means mostly stack buffer overflows and use-after-scope. However, no use-after-free yet, see below.

Tracking issue: #19.



## Why not use-after-return?

It would be really nice to also detect use-after-return bugs. However, this is difficult to do at the moment: it requires replacing stack allocations with heap allocations (there are callbacks for that: `__asan_stack_{malloc,free}`), so our ASan library would need to manage that memory.

However, right now there are *two* libraries running at the same time, PAL and LibOS, each allocating its own memory. Right now each of these has its own copy of "common" code, including `asan.c`. But since the code from both libraries runs on the same stack, the ASan library for both PAL and LibOS would need to allocate and free the same memory.

I guess one solution would be to keep only the ASan library in PAL. The `__asan_*` functions will be exported by PAL, and LibOS will not contain its own copy of the callbacks, but call the ones in PAL. Then, ASan could maintain its own structures, and use `malloc` and `free` from PAL.

I'm curious to hear your opinion on this.

## How to test this PR? 

* CI should pass
* Try triggering some stack buffer overflows manually?


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramine/240)
